File: HgStats/Services/ReviewService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using HgStats.Helpers;

namespace HgStats.Services
{
    public class ReviewService
    {
        private static readonly string authorPrefix = Guid.NewGuid().ToString();
        private const string reviewPrefix = "review:";
        private const string border = "-------";
        private const string newLine = @"\n";

        private readonly string hgRoot;
        private readonly Dictionary<string, string> authorMap;
        private readonly Dictionary<string, string> reviewerMap;
        private readonly HashSet<string> authors;
        private string AuthorMapPath => Path.Combine(hgRoot, "authormap.txt");
        private string ReviewerMapPath => Path.Combine(hgRoot, "reviewermap.txt");
        private string CD => $"cd {hgRoot}";

        public ReviewService(string hgRoot)
        {
            this.hgRoot = hgRoot;
            authorMap = GetAuthorMap();
            reviewerMap = GetReviewerMap();
            authors = GetAuthors();
        }

        private Dictionary<string, string> GetAuthorMap()
        {
            if (!File.Exists(AuthorMapPath))
                return new Dictionary<string, string>();

            return File.ReadAllLines(AuthorMapPath)
                .Select(l => l.Split('='))
                .ToDictionary(x => x[0], x => x[1]);
        }

        private Dictionary<string, string> GetReviewerMap()
        {
            if (!File.Exists(ReviewerMapPath))
                return new Dictionary<string, string>();

            return File.ReadAllLines(ReviewerMapPath)
                .Select(l => l.Split('='))
                .SelectMany(x => x[0].Split(',').Select(r => new { key = r, val = x[1] }))
                .ToDictionary(x => x.key, x => x.val);
        }

        public IEnumerable<(string author, string reviewer, int count)> GetData(string from, string to)
        {
            return GetCommits(from, to)
                .SelectMany(c => c.Reviewers.Select(r => (author : c.Author, reviewer: r)))
                .GroupBy(p => p)
                .Select(g => (author: g.Key.author, reviewer: g.Key.reviewer, count: g.Count()));
        }

        private HashSet<string> GetAuthors()
        {
            var command = $"hg log -T \"{{author}}{newLine}\"";
            var log = CmdHelper.RunViaFile(CD, command);

            return new HashSet<string>(log.Select(MapAuthor));
        }

        private List<Commit> GetCommits(string from, string to)
        {
            var dateRange = $"-d \"{from} to {to}\" ";

            var command = $"hg log {dateRange} -T \"{authorPrefix}{{author}}{newLine}{{desc}}{newLine}{border}{newLine}\"";
            var log = CmdHelper.RunViaFile(CD, command);

            var commits = new List<Commit>();
            var current = new Commit();
            foreach (var line in log)
            {
                if (line.StartsWith(authorPrefix))
                    current.Author = GetAuthor(line);

                if (line.StartsWith(reviewPrefix))
                    current.Reviewers = GetReviewers(line);

                if (line == border)
                {
                    if (current.Reviewers != null && current.Reviewers.Any())
                        commits.Add(current);
                    current = new Commit();
                }
            }

            return commits;
        }

        private string GetAuthor(string line)
        {
            var currentAuthor = line.Replace(authorPrefix, string.Empty).Trim();

            return MapAuthor(currentAuthor);
        }

        private string MapAuthor(string author)
        {
            return authorMap.ContainsKey(author)
                ? authorMap[author]
                : author;
        }

        private List<string> GetReviewers(string line)
        {
            var currentReviewers = new List<string>();
            var rawReviewers = line
                               .Replace(reviewPrefix, string.Empty)
                               .Split(new[] { ",", " " }, StringSplitOptions.RemoveEmptyEntries);
            foreach (var reviewer in rawReviewers)
            {
                if (reviewerMap.ContainsKey(reviewer))
                    currentReviewers.Add(reviewerMap[reviewer]);
                else if (authors.Contains(reviewer))
                    currentReviewers.Add(reviewer);
            }

            return currentReviewers;
        }
    }

    public class Commit
    {
        public string Author { get; set; }
        public List<string> Reviewers { get; set; }

        public override string ToString()
        {
            return $"{nameof(Author)}: {Author}, {nameof(Reviewers)}: {string.Join(", ", Reviewers)}";
        }
    }
}

File: reviewService.js
const _ = require('underscore');
const fs = require('fs');
const path = require('path');
const { runCommand } = require('./cmd.js');

const settings = require('./settings.json');

const authorPrefix = getRandomString();
const reviewPrefix = "review:";
const border = "-------";
const newLine = "\\n";
const authorMaps = {};
const authorLists = {};
const unknownReviewers = new Set();

exports.getData = (from, to) => {
    const rootResults = _.flatten(settings.hgRoots.map(root => getRootData(root, from, to)));

    const groups = _.groupBy(rootResults, c => c.author + ' + ' + c.review);
    const total = _.map(groups, g => ({
        root: 'Total',
        author: g[0].author,
        review: g[0].review,
        amount: g.reduce((acc, el) => acc + el.amount, 0)
    }));

    return total.concat(rootResults);
};

function getRootData (root, from, to) {
    const commits = getCommits(from, to, root);
    const reviewPairs = _.flatten(commits.map(c => c.review.map(r => ({author: c.author, review: r}))));
    const groups = _.groupBy(reviewPairs, c => c.author + ' + ' + c.review);

    return _.map(groups, g => ({
        root: root,
        author: g[0].author,
        review: g[0].review || g[0].author,
        amount: g.length
    }));
}

function getCommits(from, to, root) {
    initAuthorList(root);
    initAuthorMap(root);

    const dateRange = `-d \"${from.format('YYYY-MM-DD')} to ${to.format('YYYY-MM-DD')}\" `;
    const command = `hg log ${dateRange} -T \"${authorPrefix}{author}${newLine}{desc}${newLine}${border}${newLine}\"`;
    const log = runCommand(root, command);

    let commits = [];
    let current = createEmptyCommit();
    log.split('\n').forEach(line => {
        if (line.startsWith(authorPrefix))
            current.author = getAuthor(line, root);

        if (line.startsWith(reviewPrefix))
            current.review = getReview(line, root);

        if (line === border) {
            commits.push(current);
            current = createEmptyCommit();
        }
    });

    return commits;
}

function initAuthorMap(root) {
    if (authorMaps[root])
        return;

    const mapPath = path.join(root, 'authormap.txt');
    if (!fs.existsSync(mapPath))
        return;

    authorMaps[root] = [];
    fs.readFileSync(mapPath)
        .toString()
        .split(/[\r\n]/)
        .filter(l => l)
        .forEach(line => {
            const tokens = line.split('=');
            authorMaps[root][tokens[0]] = tokens[1];
            authorLists[root].add(tokens[1]);
        });
}

function initAuthorList(root) {
    const command = `hg log -T \"{author}${newLine}\"`;
    const log = runCommand(root, command);
    authorLists[root] = new Set(log.split('\n').filter(l => l));
}

function getAuthor(line, root) {
    let author = line.replace(authorPrefix, '').trim();
    return mapAuthor(author, root);
}

function getReview(line, root) {
    const result = [];
    line.replace(reviewPrefix, '')
        .split(/[\s,]+/)
        .filter(t => t)
        .map(author => mapAuthor(author, root))
        .forEach(author => authorLists[root].has(author)
            ? result.push(author)
            : unknownReviewers.add(author));
    return result;
}

function mapAuthor(author, root) {
    const map = authorMaps[root];
    return map
        ? map[author] || author
        : author;
}

function createEmptyCommit() {
    return {review: []};
}

function getRandomString() {
    return Math.random().toString(36).substr(2);
}
